packages/loopring_v3/contracts/impl/libexchange/ExchangeWithdrawals.sol

```diff
@@ -37,317 +22,255 @@ library ExchangeWithdrawals
-        address token
-        )
-        public
-    {
-        uint24 accountID = (owner != address(0)) ? S.getAccountID(owner) : 0;
-        uint16 tokenID = S.getTokenID(token);
-        uint amount = S.amountWithdrawable[accountID][tokenID];
+        uint amount = deposit.amount;
+        uint fee = deposit.fee;
 
-        // Make sure this amount can't be withdrawn again
-        S.amountWithdrawable[accountID][tokenID] = 0;
+        // Reset the deposit request
+        S.pendingDeposits[owner][tokenID][index] = ExchangeData.Deposit(0, 0, 0);
 
         // Transfer the tokens
         transferTokens(
             S,
-            accountID,
+            owner,
+            owner,
             tokenID,
             amount,
+            new bytes(0),
+            gasleft(),
             false
         );
+
+        // Return the fee
+        owner.sendETHAndVerify(fee, gasleft());
     }
 
-    function distributeWithdrawals(
+    function withdrawFromApprovedWithdrawals(
         ExchangeData.State storage S,
-        bytes memory withdrawals
+        address[] memory owners,
+        address[] memory tokens
         )
         public
     {
-        for (uint i = 0; i < withdrawals.length; i += 8) {
-            uint data = uint(withdrawals.bytesToUint64(i));
+        require(owners.length == tokens.length, "INVALID_INPUT_DATA");
+        for (uint i = 0; i < owners.length; i++) {
+            address owner = owners[i];
+            uint16 tokenID = S.getTokenID(tokens[i]);
+            uint amount = S.amountWithdrawable[owner][tokenID];
 
-            // Extract the withdrawal data
-            uint16 tokenID = uint16((data >> 48) & 0xFFFF);
-            uint24 accountID = uint24((data >> 24) & 0xFFFFFF);
-            uint amount = (data & 0xFFFFFF).decodeFloat(24);
+            // Make sure this amount can't be withdrawn again
+            S.amountWithdrawable[owner][tokenID] = 0;
 
             // Transfer the tokens
-            bool success = transferTokens(
+            transferTokens(
                 S,
-                accountID,
+                owner,
+                owner,
                 tokenID,
                 amount,
-                true
+                new bytes(0),
+                gasleft(),
+                false
             );
-            if (!success) {
-                // Allow the amount to be withdrawn using `withdrawFromApprovedWithdrawal`.
-                S.amountWithdrawable[accountID][tokenID] = S.amountWithdrawable[accountID][tokenID].add(amount);
-            }
         }
     }
 
-    // == Internal and Private Functions ==
-
-    function getOwner(
+    function distributeWithdrawal(
         ExchangeData.State storage S,
-        uint24 accountID
+        address from,
+        address to,
+        uint16  tokenID,
+        uint    amount,
+        bytes   memory auxiliaryData,
+        uint    gasLimit
         )
-        private
-        view
-        returns (address addr)
+        public
     {
```

Comment:
Do we need to make sure S.amountWithdrawable[to][tokenID] contains at least amount before real transferTokens? Right now we have no such check, looks like any one can distributeWithdrawl money even not approved.

Reply:
This is a `public` function on a library contract, it cannot be called directly and needs to pass through the exchange functions (well, it can be called directly but it will work on a different contract state than when called from the exchange contract because calls to library functions use `delegatecall` internally).

Reply:
> This is a `public` function on a library contract, it cannot be called directly and needs to pass through the exchange functions (well, it can be called directly but it will work on a different contract state than when called from the exchange contract because calls to library functions use `delegatecall` internally).

Okay, then we'd better to add one test case for this behavior. And make sure cannot transfer any asset if user directly called instead of delegatecall from exchange.

Reply:
Indirectly we already test this because only the exchange contract can interact with the deposit contract (tested) and only the exchange contract can modify its own contract state (let's hope we can assume that, otherwise smart contracts don't work). But I guess an extra test to verify this can't hurt, though I don't even know how to correctly call a function using a `storage` variable because it's some kind of pointer...